### ISSUE TITLE
Add string charoffset-to-byteoffset index to large strings to speed up UTF8 scan

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#335]
Currently random access to a character at a specified index is dependent on a heap level string cache which keeps of a few strings and their last used charoffset-to-byteoffset mappings. The string cache allows a lookup to proceed by scanning backwards or forwards from a previously used index so that e.g. character loops are fast.

This is fine for low memory devices, but doesn't perform well in all cases, and has an unpredictable performance profile. In the worst case an `O(n)` algorithm (where `n` is string length) can become `O(n^2)`.

One simple improvement would be to add a small index for larger strings, e.g. provide a charoffset-to-byteoffset mapping for every 64 character offsets so that scanning would at most need to go backwards or forwards 32 characters. The cost of such an index would be 4 bytes per 64 codepoints (at most ~6-7%) and would make the lookup performance more predictable and would eliminate worst case behaviors like `O(n^2)` string scanning.

This index is probably not good for low memory devices because it requires one more pointer per string. But it should be enabled by default, and the string cache can either be disabled or made smaller by default.

**Comment by fatcerberus** [comment]
Is there a lot of real-world code that jumps around strings like this?  With the current setup it seems the only way to trigger the O(n^2) behavior, would be either:

* The characters aren't processed linearly in order (either forwards or backwards).
* Multiple strings are involved in the processing causing cache thrashing.

**Comment by svaarala** [comment]
It's hard to predict because there's a lot of code base on e.g. web pages. For example, one can use strings as 16-bit lookup tables in which case their usage might be quite unexpected - random accesses might be more normal than in other cases.

The general expectation there would be that it's not necessarily fast, but it's still `O(1)` with some constant and it's surprising if that's not the case. It's a nice property of an engine if performance behavior is somehow "continuous" to the input, so that there are minimal cases where performance suddenly drops a great deal.

**Comment by fatcerberus** [comment]
Agreed on the predictable performance, I was mostly just curious how much of a real-world issue it is in practice.

**Comment by svaarala** [comment]
Yeah, don't really know the answer to that :) This is more of a potential offender than an actual any. #334 was a concrete problem.
